src/cmds/qstat.c

```diff
@@ -302,49 +358,153 @@ states(char *string, char *q, char *r, char *h, char *w, char *t, char *e, int l
+	char           *comma = ",";
+	static char    *prev_resc_name = NULL;
+	char           *buf = NULL;
+	char           **keyvalpair = NULL;
+	char           *key = NULL;
+	char           *val = NULL;
+	char           *tokptr = NULL;
+
+	if(v == NULL)
+		return;
+	switch (output_format) {
+		case FORMAT_JSON:
+			if (strcmp(n, ATTR_v) == 0) {
+				if (add_json_node(JSON_OBJECT, JSON_NULL, n, NULL) == NULL) {
+					fprintf(stderr, "qstat: out of memory\n");
+					exit(1);
+				}
+				keyvalpair = break_comma_list(v);
+				for(i = 0; keyvalpair[i]; i++)
+				{
+					key = string_token(keyvalpair[i], "=", &tokptr);
+					val = string_token(NULL, "=", &tokptr);
+					if(add_json_node(JSON_VALUE, JSON_NULL, key, val) == NULL) {
+						fprintf(stderr, "qstat: out of memory\n");
+						exit(1);
+					}
+				}
+				free_string_array(keyvalpair);
+				if (add_json_node(JSON_OBJECT_END, JSON_NULL, NULL, NULL) == NULL) {
+					fprintf(stderr, "qstat: out of memory\n");
+					exit(1);
+				}
 			}
-			while (*c) {
-				putchar(*c++);
-				if (++start > 78) {
-					start = 8;
-					printf("\n\t");
+			else {
+				if(r) {
+					if (prev_resc_name && strcmp(prev_resc_name, n) != 0) {
+						if (add_json_node(JSON_OBJECT_END, JSON_NULL, NULL, NULL) == NULL) {
+							fprintf(stderr, "qstat: out of memory\n");
+							exit(1);
+						}
+						prev_resc_name = NULL;
+					}
+					if(prev_resc_name == NULL || strcmp(prev_resc_name, n) != 0) {
+						if (add_json_node(JSON_OBJECT, JSON_NULL, n, NULL) == NULL) {
+							fprintf(stderr, "qstat: out of memory\n");
+							exit(1);
+						}
+						prev_resc_name = n;
+					}
+					if(add_json_node(JSON_VALUE, JSON_NULL, r, v) == NULL) {
+						fprintf(stderr, "qstat: out of memory\n");
+						exit(1);
+					}
+				}
+				else {
+					if(prev_resc_name) {
+						if (add_json_node(JSON_OBJECT_END, JSON_NULL, NULL, NULL) == NULL) {
+							fprintf(stderr, "qstat: out of memory\n");
+							exit(1);
+					}
+					prev_resc_name = NULL;
+					}
+					if(add_json_node(JSON_VALUE, JSON_NULL, n, v) == NULL) {
+					fprintf(stderr, "qstat: out of memory\n");
+					exit(1);
+					}
+				}
+			}
+			break;
+
+		case FORMAT_DSV:
+			buf = escape_delimiter(v, delimiter, '\\');
+			if(buf == NULL) {
+				fprintf(stderr, "qstat: out of memory\n");
+				exit(1);
+			}
+			buf = escape_feed_chars(buf);
+			if(buf == NULL) {
+				fprintf(stderr, "qstat: out of memory\n");
+				exit(1);
+			}
+			if (r)
+				printf("%s.%s=%s", n, r, buf);
+			else
+				printf("%s=%s", n, buf);
+			free(buf);
+			break;
+
+		default:
+			if(one_line) {
+				buf = escape_feed_chars(v);
+				if(buf == NULL) {
+					fprintf(stderr, "qstat: out of memory\n");
+					exit(1);
+				}
+				if (r)
+					printf("    %s.%s = %s", n, r, buf);
+				else
+					printf("    %s = %s", n, buf);
+				free(buf);
+			}
+			else {
+				start = strlen(n) + 7;	/* 4 spaces + ' = ' is 7 */
+				printf("    %s", n);
+				if (r) {
+					start += strlen(r) + 1;
```

Comment:
Line 475 to 505 is legacy code. I'll try to add more comments and macros to make the code more readable.